File: TRPO/lab6/project-root/bin/main/IntegerType.kt
import java.io.InputStream

class IntegerType : UserType {
    override fun typeName() = "Integer"

    override fun create() = 0

    override fun clone(obj: Any) = (obj as Int)

    override fun readValue(input: InputStream) = 
        parseValue(input.bufferedReader().readLine())

    override fun parseValue(ss: String) = ss.toInt()

    override fun getComparator() = object : Comparator {
        override fun compare(o1: Any?, o2: Any?) = 
            (o1 as Int).compareTo(o2 as Int)
    }
}

File: TRPO/lab5/scala_lab5/src/main/scala/types/IntegerType.scala
// package types

import java.io.InputStreamReader
import java.io.BufferedReader

class IntegerType extends UserType {
  override def typeName: String = "Integer"

  override def create: Any = 0

  override def clone(obj: Any): Any = obj match {
    case i: Int => i 
    case _ => throw new IllegalArgumentException("Invalid Integer object")
  }

  override def readValue(in: InputStreamReader): Any = {
    val br = new BufferedReader(in)
    parseValue(br.readLine())
  }

  override def parseValue(ss: String): Any = ss.toInt

  override def getTypeComparator: Comparator = new Comparator {
    override def compare(o1: Any, o2: Any): Int = 
      o1.asInstanceOf[Int] - o2.asInstanceOf[Int]
  }
}